<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# Exploring and pre-processing a dataset using Pandas 


Estimated time needed: **30** minutes
    

## Objectives

After completing this lab you will be able to:

* Explore the dataset
* Pre-process dataset as required (may be for visualization)


## Introduction

The aim of this lab is to provide you a refresher on the **Pandas** library, so that you can pre-process and anlyse the datasets before applying data visualization techniques on it. This lab will work as acrash course on *pandas*. if you are interested in learning more about the *pandas* library, detailed description and explanation of how to use it and how to clean, munge, and process data stored in a *pandas* dataframe are provided in other IBM courses.

------------


## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

1. [Exploring Datasets with *pandas*](#0)<br>
2. [The Dataset: Immigration to Canada from 1980 to 2013](#1)<br>
3. [*pandas* Basics](#2) <br>
4. [*pandas* Intermediate: Indexing and Selection](#3) <br>
5. [*pandas* Filtering based on a criteria](#4)<br>
6. [*pandas* Sorting Values](#5)

</div>


# Exploring Datasets with *pandas* <a id="0"></a>

*pandas* is an essential data analysis toolkit for Python. From their [website](http://pandas.pydata.org/):
>*pandas* is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. It aims to be the fundamental high-level building block for doing practical, **real world** data analysis in Python.

The course heavily relies on *pandas* for data wrangling, analysis, and visualization. We encourage you to spend some time and familiarize yourself with the *pandas* API Reference: http://pandas.pydata.org/pandas-docs/stable/api.html.


## The Dataset: Immigration to Canada from 1980 to 2013 <a id="1"></a>


Dataset Source: [International migration flows to and from selected countries - The 2015 revision](https://www.un.org/development/desa/pd/data/international-migration-flows).

The dataset contains annual data on the flows of international immigrants as recorded by the countries of destination. The data presents both inflows and outflows according to the place of birth, citizenship or place of previous / next residence both for foreigners and nationals. The current version presents data pertaining to 45 countries.

In this lab, we will focus on the Canadian immigration data.

![Data Preview](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/Module%201/images/DataSnapshot.png)

 The Canada Immigration dataset can be fetched from <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx">here</a>.

---


## *pandas* Basics<a id="2"></a>


The first thing we'll do is install **openpyxl** (formerly **xlrd**), a module that *pandas* requires to read Excel files.


In [ ]:
!mamba install openpyxl==3.0.9 -y

Next, we'll do is import two key data analysis modules: *pandas* and *numpy*.


In [ ]:
import numpy as np  # useful for many scientific computing in Python
import pandas as pd # primary data structure library

In [1]:
import numpy as np
import pandas as pd

Let's download and import our primary Canadian Immigration dataset using *pandas*'s `read_excel()` method.


In [ ]:
df_can = pd.read_excel(
    'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx',
    sheet_name='Canada by Citizenship',
    skiprows=range(20),
    skipfooter=2)

print('Data read into a pandas dataframe!')

In [2]:
df_can = pd.read_excel(
    'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx',
    sheet_name ='Canada by Citizenship',
    skiprows=range(20),
    skipfooter=2)
print('Data read into a pandas dataframe!')

Data read into a pandas dataframe!


Let's view the top 5 rows of the dataset using the `head()` function.


In [3]:
df_can.head()
# tip: You can specify the number of rows you'd like to see as follows: df_can.head(10) 

Type    Coverage          OdName  AREA AreaName   REG  \
0  Immigrants  Foreigners     Afghanistan   935     Asia  5501   
1  Immigrants  Foreigners         Albania   908   Europe   925   
2  Immigrants  Foreigners         Algeria   903   Africa   912   
3  Immigrants  Foreigners  American Samoa   909  Oceania   957   
4  Immigrants  Foreigners         Andorra   908   Europe   925   

           RegName  DEV             DevName  1980  1981  1982  1983  1984  \
0    Southern Asia  902  Developing regions    16    39    39    47    71   
1  Southern Europe  901   Developed regions     1     0     0     0     0   
2  Northern Africa  902  Developing regions    80    67    71    69    63   
3        Polynesia  902  Developing regions     0     1     0     0     0   
4  Southern Europe  901   Developed regions     0     0     0     0     0   

   1985  1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  1996  \
0   340   496   741   828  1076  1028  1378  1170   713   858  1537  2212   
1     0     1     2     2     3     3    21    56    96    71    63   113   
2    44    69   132   242   434   491   872   795   717   595  1106  2054   
3     0     0     1     0     1     2     0     0     0     0     0     0   
4     0     2     0     0     0     3     0     1     0     0     0     0   

   1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  2008  \
0  2555  1999  2395  3326  4067  3697  3479  2978  3436  3009  2652  2111   
1   307   574  1264  1816  1602  1021   853  1450  1223   856   702   560   
2  1842  2292  2389  2867  3418  3406  3072  3616  3626  4807  3623  4005   
3     0     0     0     0     0     0     0     0     0     1     0     0   
4     0     2     0     0     1     0     2     0     0     1     1     0   

   2009  2010  2011  2012  2013  
0  1746  1758  2203  2635  2004  
1   716   561   539   620   603  
2  5393  4752  4325  3774  4331  
3     0     0     0     0     0  
4     0     0     0     1     1

We can also view the bottom 5 rows of the dataset using the `tail()` function.


In [4]:
df_can.tail()

Type    Coverage          OdName  AREA AreaName  REG  \
190  Immigrants  Foreigners        Viet Nam   935     Asia  920   
191  Immigrants  Foreigners  Western Sahara   903   Africa  912   
192  Immigrants  Foreigners           Yemen   935     Asia  922   
193  Immigrants  Foreigners          Zambia   903   Africa  910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa  910   

                RegName  DEV             DevName  1980  1981  1982  1983  \
190  South-Eastern Asia  902  Developing regions  1191  1829  2162  3404   
191     Northern Africa  902  Developing regions     0     0     0     0   
192        Western Asia  902  Developing regions     1     2     1     6   
193      Eastern Africa  902  Developing regions    11    17    11     7   
194      Eastern Africa  902  Developing regions    72   114   102    44   

     1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  \
190  7583  5907  2741  1406  1411  3004  3801  5870  5416  6547  5105  3723   
191     0     0     0     0     0     0     0     1     0     0     0     0   
192     0    18     7    12     7    18     4    18    41    41    39    73   
193    16     9    15    23    44    68    77    69    73    46    51    41   
194    32    29    43    68    99   187   129    94    61    72    78    58   

     1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  \
190  2462  1752  1631  1419  1803  2117  2291  1713  1816  1852  3153  2574   
191     0     0     0     0     0     0     0     0     0     0     1     0   
192   144   121   141   134   122   181   171   113   124   161   140   122   
193    34    72    34    51    39    78    50    46    56    91    77    71   
194    39    44    43    49    98   110   191   669  1450   615   454   663   

     2008  2009  2010  2011  2012  2013  
190  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0  
192   133   128   211   160   174   217  
193    64    60   102    69    46    59  
194   611   508   494   434   437   407

When analyzing a dataset, it's always a good idea to start by getting basic information about your dataframe. We can do this by using the `info()` method.

This method can be used to get a short summary of the dataframe.


In [6]:
df_can.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Columns: 43 entries, Type to 2013
dtypes: int64(37), object(6)
memory usage: 65.6+ KB


In [7]:
df_can.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Columns: 43 entries, Type to 2013
dtypes: int64(37), object(6)
memory usage: 65.6+ KB


To get the list of column headers we can call upon the data frame's `columns` instance variable.


In [ ]:
df_can.columns

In [8]:
df_can.columns

Index([    'Type', 'Coverage',   'OdName',     'AREA', 'AreaName',      'REG',
        'RegName',      'DEV',  'DevName',       1980,       1981,       1982,
             1983,       1984,       1985,       1986,       1987,       1988,
             1989,       1990,       1991,       1992,       1993,       1994,
             1995,       1996,       1997,       1998,       1999,       2000,
             2001,       2002,       2003,       2004,       2005,       2006,
             2007,       2008,       2009,       2010,       2011,       2012,
             2013],
      dtype='object')

Similarly, to get the list of indices we use the `.index` instance variables.


In [9]:
df_can.index

RangeIndex(start=0, stop=195, step=1)

In [ ]:
df_can.index

Note: The default type of intance variables `index` and `columns` are **NOT** `list`.


In [10]:
print(type(df_can.columns))
print(type(df_can.index))

<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


To get the index and columns as lists, we can use the `tolist()` method.


In [ ]:
df_can.columns.tolist()

In [11]:
df_can.columns.tolist()

['Type',
 'Coverage',
 'OdName',
 'AREA',
 'AreaName',
 'REG',
 'RegName',
 'DEV',
 'DevName',
 1980,
 1981,
 1982,
 1983,
 1984,
 1985,
 1986,
 1987,
 1988,
 1989,
 1990,
 1991,
 1992,
 1993,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013]

In [ ]:
df_can.index.tolist()

In [12]:
df_can.index.tolist()

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


In [13]:
print(type(df_can.columns.tolist()))
print(type(df_can.index.tolist()))

<class 'list'>
<class 'list'>


To view the dimensions of the dataframe, we use the `shape` instance variable of it.


In [14]:
df_can.shape

(195, 43)

In [ ]:
# size of dataframe (rows, columns)
df_can.shape

**Note**: The main types stored in *pandas* objects are `float`, `int`, `bool`, `datetime64[ns]`, `datetime64[ns, tz]`, `timedelta[ns]`, `category`, and `object` (string). In addition, these dtypes have item sizes, e.g. `int64` and `int32`.


Let's clean the data set to remove a few unnecessary columns. We can use *pandas* `drop()` method as follows:


In [ ]:
# in pandas axis=0 represents rows (default) and axis=1 represents columns.
df_can.drop(['AREA','REG','DEV','Type','Coverage'], axis=1, inplace=True)
df_can.head(2)

In [15]:
df_can.drop(['AREA','REG','DEV','Type','Coverage'],axis=1,inplace=True)
df_can.head()

OdName AreaName          RegName             DevName  1980  1981  \
0     Afghanistan     Asia    Southern Asia  Developing regions    16    39   
1         Albania   Europe  Southern Europe   Developed regions     1     0   
2         Algeria   Africa  Northern Africa  Developing regions    80    67   
3  American Samoa  Oceania        Polynesia  Developing regions     0     1   
4         Andorra   Europe  Southern Europe   Developed regions     0     0   

   1982  1983  1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  \
0    39    47    71   340   496   741   828  1076  1028  1378  1170   713   
1     0     0     0     0     1     2     2     3     3    21    56    96   
2    71    69    63    44    69   132   242   434   491   872   795   717   
3     0     0     0     0     0     1     0     1     2     0     0     0   
4     0     0     0     0     2     0     0     0     3     0     1     0   

   1994  1995  1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  \
0   858  1537  2212  2555  1999  2395  3326  4067  3697  3479  2978  3436   
1    71    63   113   307   574  1264  1816  1602  1021   853  1450  1223   
2   595  1106  2054  1842  2292  2389  2867  3418  3406  3072  3616  3626   
3     0     0     0     0     0     0     0     0     0     0     0     0   
4     0     0     0     0     2     0     0     1     0     2     0     0   

   2006  2007  2008  2009  2010  2011  2012  2013  
0  3009  2652  2111  1746  1758  2203  2635  2004  
1   856   702   560   716   561   539   620   603  
2  4807  3623  4005  5393  4752  4325  3774  4331  
3     1     0     0     0     0     0     0     0  
4     1     1     0     0     0     0     1     1

Let's rename the columns so that they make sense. We can use `rename()` method by passing in a dictionary of old and new names as follows:


In [ ]:
df_can.rename(columns={'OdName':'Country', 'AreaName':'Continent', 'RegName':'Region'}, inplace=True)
df_can.columns

In [17]:
df_can.rename(columns={'OdName':'Country', 'AreaName':'Continent', 'RegName':'Region'}, inplace=True)
df_can.columns

Index([  'Country', 'Continent',    'Region',   'DevName',        1980,
              1981,        1982,        1983,        1984,        1985,
              1986,        1987,        1988,        1989,        1990,
              1991,        1992,        1993,        1994,        1995,
              1996,        1997,        1998,        1999,        2000,
              2001,        2002,        2003,        2004,        2005,
              2006,        2007,        2008,        2009,        2010,
              2011,        2012,        2013],
      dtype='object')

We will also add a 'Total' column that sums up the total immigrants by country over the entire period 1980 - 2013, as follows:


In [20]:
df_can['Total']=df_can.sum(numeric_only=True,axis=1)
df_can['Total']

0       117278
1        31398
2       138878
3           12
4           30
5         4226
6         1962
7        39192
8         6620
9        47658
10        9970
11        5298
12        1616
13         950
14      131136
15       13846
16       15950
17       21524
18        2166
19        5704
20       11752
21        6410
22       43710
23         792
24       59318
25        1200
26       46620
27        4086
28       16220
29         402
30       13076
31       32042
32          40
33        1100
34        3306
35       42718
36     1319924
37       18654
38         568
39      144176
40         388
41        3372
42        8978
43       16582
44       16002
45       40730
46        2252
47        8384
48         776
49       51164
50        7802
51        3292
52        4940
53       22386
54       24466
55      145490
56      109756
57         122
58       20900
59        1870
60       70272
61       41218
62        5562
63      218182
64        1904
65        1206
66        

In [ ]:
df_can['Total'] = df_can.sum(axis=1)
df_can['Total']

We can check to see how many null objects we have in the dataset as follows:


In [ ]:
df_can.isnull().sum()

In [21]:
df_can.isnull().sum()

Country      0
Continent    0
Region       0
DevName      0
1980         0
1981         0
1982         0
1983         0
1984         0
1985         0
1986         0
1987         0
1988         0
1989         0
1990         0
1991         0
1992         0
1993         0
1994         0
1995         0
1996         0
1997         0
1998         0
1999         0
2000         0
2001         0
2002         0
2003         0
2004         0
2005         0
2006         0
2007         0
2008         0
2009         0
2010         0
2011         0
2012         0
2013         0
Total        0
dtype: int64

Finally, let's view a quick summary of each column in our dataframe using the `describe()` method.


In [ ]:
df_can.describe()

In [22]:
df_can.describe()

1980          1981          1982          1983          1984  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     508.394872    566.989744    534.723077    387.435897    376.497436   
std     1949.588546   2152.643752   1866.997511   1204.333597   1198.246371   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%       13.000000     10.000000     11.000000     12.000000     13.000000   
75%      251.500000    295.500000    275.000000    173.000000    181.000000   
max    22045.000000  24796.000000  20620.000000  10015.000000  10170.000000   

              1985         1986          1987          1988          1989  \
count   195.000000   195.000000    195.000000    195.000000    195.000000   
mean    358.861538   441.271795    691.133333    714.389744    843.241026   
std    1079.309600  1225.576630   2109.205607   2443.606788   2555.048874   
min       0.000000     0.000000      0.000000      0.000000      0.000000   
25%       0.000000     0.500000      0.500000      1.000000      1.000000   
50%      17.000000    18.000000     26.000000     34.000000     44.000000   
75%     197.000000   254.000000    434.000000    409.000000    508.500000   
max    9564.000000  9470.000000  21337.000000  27359.000000  23795.000000   

               1990          1991          1992          1993          1994  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     964.379487   1064.148718   1136.856410   1138.712821    993.153846   
std     3158.730195   2952.093731   3330.083742   3495.220063   3613.336444   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        2.000000      3.000000      6.500000     11.500000      9.500000   
50%       38.000000     51.000000     74.000000     85.000000     76.000000   
75%      612.500000    657.500000    655.000000    722.500000    545.000000   
max    31668.000000  23380.000000  34123.000000  33720.000000  39231.000000   

               1995          1996          1997          1998          1999  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     962.625641   1026.076923    989.153846    824.241026    922.143590   
std     3091.492343   3321.045004   3070.761447   2385.943695   2887.632585   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       10.500000     14.500000     19.500000     15.000000     16.000000   
50%       91.000000    118.000000    114.000000    106.000000    116.000000   
75%      550.500000    603.500000    612.500000    535.500000    548.500000   
max    30145.000000  29322.000000  22965.000000  21049.000000  30069.000000   

               2000          2001          2002          2003          2004  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1111.343590   1244.323077   1144.158974   1114.343590   1190.169231   
std     3664.042361   3961.621410   3660.579836   3623.509519   3710.505369   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       16.000000     22.000000     18.500000     21.500000     19.000000   
50%      138.000000    169.000000    165.000000    161.000000    191.000000   
75%      659.000000    793.500000    686.000000    673.500000    756.500000   
max    35529.000000  36434.000000  31961.000000  36439.000000  36619.000000   

               2005          2006          2007          2008          2009  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1320.292308   1266.958974   1191.820513   1246.394872   1275.733333   
std     4425.957828   3926.717747   3443.542409   3694.573544   3829.630424   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       28.500000     25.000000     31.000000     31.000000     36.000000   

---
## *pandas* Intermediate: Indexing and Selection (slicing)<a id="3"></a>


### Select Column
**There are two ways to filter on a column name:**

Method 1: Quick and easy, but only works if the column name does NOT have spaces or special characters.
```python
    df.column_name               # returns series
```

Method 2: More robust, and can filter on multiple columns.

```python
    df['column']                  # returns series
```

```python 
    df[['column 1', 'column 2']]  # returns dataframe
```
---


Example: Let's try filtering on the list of countries ('Country').


In [23]:
df_can.Country

0                                               Afghanistan
1                                                   Albania
2                                                   Algeria
3                                            American Samoa
4                                                   Andorra
5                                                    Angola
6                                       Antigua and Barbuda
7                                                 Argentina
8                                                   Armenia
9                                                 Australia
10                                                  Austria
11                                               Azerbaijan
12                                                  Bahamas
13                                                  Bahrain
14                                               Bangladesh
15                                                 Barbados
16                                      

In [ ]:
df_can.Country  # returns a series

Let's try filtering on the list of countries ('Country') and the data for years: 1980 - 1985.


In [24]:
df_can[['Country',1980,1981,1982,1983,1984,1985]]

Country   1980   1981  \
0                                             Afghanistan     16     39   
1                                                 Albania      1      0   
2                                                 Algeria     80     67   
3                                          American Samoa      0      1   
4                                                 Andorra      0      0   
5                                                  Angola      1      3   
6                                     Antigua and Barbuda      0      0   
7                                               Argentina    368    426   
8                                                 Armenia      0      0   
9                                               Australia    702    639   
10                                                Austria    234    238   
11                                             Azerbaijan      0      0   
12                                                Bahamas     26     23   
13                                                Bahrain      0      2   
14                                             Bangladesh     83     84   
15                                               Barbados    372    376   
16                                                Belarus      0      0   
17                                                Belgium    511    540   
18                                                 Belize     16     27   
19                                                  Benin      2      5   
20                                                 Bhutan      0      0   
21                       Bolivia (Plurinational State of)     44     52   
22                                 Bosnia and Herzegovina      0      0   
23                                               Botswana     10      1   
24                                                 Brazil    211    220   
25                                      Brunei Darussalam     79      6   
26                                               Bulgaria     24     20   
27                                           Burkina Faso      2      1   
28                                                Burundi      0      0   
29                                             Cabo Verde      1      1   
30                                               Cambodia     12     19   
31                                               Cameroon      9      2   
32                                                 Canada      0      0   
33                               Central African Republic      4      3   
34                                                   Chad      0      0   
35                                                  Chile   1233   1069   
36                                                  China   5123   6682   
37         China, Hong Kong Special Administrative Region      0      0   
38             China, Macao Special Administrative Region      0      0   
39                                               Colombia    266    326   
40                                                Comoros      0      2   
41                                                  Congo      2      0   
42                                             Costa Rica     28     35   
43                                          Côte d'Ivoire     14      8   
44                                                Croatia      0      0   
45                                                   Cuba    306     41   
46                                                 Cyprus    132    128   
47                                         Czech Republic      0      0   
48                  Democratic People's Republic of Korea      1      1   
49                       Democratic Republic of the Congo     15     16   
50                                                Denmark    272    293   
51                                               Djibouti      0      1   
52                                               Dominica    104     81   


In [ ]:
df_can[['Country', 1980, 1981, 1982, 1983, 1984, 1985]] # returns a dataframe
# notice that 'Country' is string, and the years are integers. 
# for the sake of consistency, we will convert all column names to string later on.

### Select Row

There are main 2 ways to select rows:

```python
    df.loc[label]    # filters by the labels of the index/column
    df.iloc[index]   # filters by the positions of the index/column
```


Before we proceed, notice that the default index of the dataset is a numeric range from 0 to 194. This makes it very difficult to do a query by a specific country. For example to search for data on Japan, we need to know the corresponding index value.

This can be fixed very easily by setting the 'Country' column as the index using `set_index()` method.


In [ ]:
df_can.set_index('Country', inplace=True)
# tip: The opposite of set is reset. So to reset the index, we can use df_can.reset_index()

In [25]:
df_can.set_index('Country',inplace=True)

In [26]:
df_can.head(3)

Continent           Region             DevName  1980  1981  1982  \
Country                                                                        
Afghanistan      Asia    Southern Asia  Developing regions    16    39    39   
Albania        Europe  Southern Europe   Developed regions     1     0     0   
Algeria        Africa  Northern Africa  Developing regions    80    67    71   

             1983  1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  \
Country                                                                         
Afghanistan    47    71   340   496   741   828  1076  1028  1378  1170   713   
Albania         0     0     0     1     2     2     3     3    21    56    96   
Algeria        69    63    44    69   132   242   434   491   872   795   717   

             1994  1995  1996  1997  1998  1999  2000  2001  2002  2003  2004  \
Country                                                                         
Afghanistan   858  1537  2212  2555  1999  2395  3326  4067  3697  3479  2978   
Albania        71    63   113   307   574  1264  1816  1602  1021   853  1450   
Algeria       595  1106  2054  1842  2292  2389  2867  3418  3406  3072  3616   

             2005  2006  2007  2008  2009  2010  2011  2012  2013   Total  
Country                                                                    
Afghanistan  3436  3009  2652  2111  1746  1758  2203  2635  2004  117278  
Albania      1223   856   702   560   716   561   539   620   603   31398  
Algeria      3626  4807  3623  4005  5393  4752  4325  3774  4331  138878

In [29]:
df_can.loc[['Japan']]

Continent        Region            DevName  1980  1981  1982  1983  \
Country                                                                      
Japan        Asia  Eastern Asia  Developed regions   701   756   598   309   

         1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  1994  \
Country                                                                     
Japan     246   198   248   422   324   494   379   506   605   907   956   

         1995  1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  \
Country                                                                     
Japan     826   994   924   897  1083  1010  1092   806   817   973  1067   

         2006  2007  2008  2009  2010  2011  2012  2013  Total  
Country                                                         
Japan    1212  1250  1284  1194  1168  1265  1214   982  55414

In [34]:
# optional: to remove the name of the index
df_can.index.name = None

In [32]:
df_can.index.name=None

In [35]:
df_can.head()

Continent           Region             DevName  1980  1981  \
Afghanistan         Asia    Southern Asia  Developing regions    16    39   
Albania           Europe  Southern Europe   Developed regions     1     0   
Algeria           Africa  Northern Africa  Developing regions    80    67   
American Samoa   Oceania        Polynesia  Developing regions     0     1   
Andorra           Europe  Southern Europe   Developed regions     0     0   

                1982  1983  1984  1985  1986  1987  1988  1989  1990  1991  \
Afghanistan       39    47    71   340   496   741   828  1076  1028  1378   
Albania            0     0     0     0     1     2     2     3     3    21   
Algeria           71    69    63    44    69   132   242   434   491   872   
American Samoa     0     0     0     0     0     1     0     1     2     0   
Andorra            0     0     0     0     2     0     0     0     3     0   

                1992  1993  1994  1995  1996  1997  1998  1999  2000  2001  \
Afghanistan     1170   713   858  1537  2212  2555  1999  2395  3326  4067   
Albania           56    96    71    63   113   307   574  1264  1816  1602   
Algeria          795   717   595  1106  2054  1842  2292  2389  2867  3418   
American Samoa     0     0     0     0     0     0     0     0     0     0   
Andorra            1     0     0     0     0     0     2     0     0     1   

                2002  2003  2004  2005  2006  2007  2008  2009  2010  2011  \
Afghanistan     3697  3479  2978  3436  3009  2652  2111  1746  1758  2203   
Albania         1021   853  1450  1223   856   702   560   716   561   539   
Algeria         3406  3072  3616  3626  4807  3623  4005  5393  4752  4325   
American Samoa     0     0     0     0     1     0     0     0     0     0   
Andorra            0     2     0     0     1     1     0     0     0     0   

                2012  2013   Total  
Afghanistan     2635  2004  117278  
Albania          620   603   31398  
Algeria         3774  4331  138878  
American Samoa     0     0      12  
Andorra            1     1      30

Example: Let's view the number of immigrants from Japan (row 87) for the following scenarios:
    1. The full row data (all columns)
    2. For year 2013
    3. For years 1980 to 1985


In [ ]:
# 1. the full row data (all columns)
df_can.loc['Japan']

In [36]:
df_can.loc['Japan']

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [37]:
# alternate methods
df_can.iloc[87]

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [ ]:
df_can[df_can.index == 'Japan']

In [38]:
df_can[df_can.index=='Japan']

Continent        Region            DevName  1980  1981  1982  1983  \
Japan      Asia  Eastern Asia  Developed regions   701   756   598   309   

       1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  \
Japan   246   198   248   422   324   494   379   506   605   907   956   826   

       1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  \
Japan   994   924   897  1083  1010  1092   806   817   973  1067  1212  1250   

       2008  2009  2010  2011  2012  2013  Total  
Japan  1284  1194  1168  1265  1214   982  55414

In [ ]:
# 2. for year 2013
df_can.loc['Japan', 2013]

In [40]:
df_can.loc['Japan',2013]

np.int64(982)

In [ ]:
# alternate method
# year 2013 is the last column, with a positional index of 36
df_can.iloc[87, 36]

In [41]:
df_can.iloc[87,36]

np.int64(982)

In [ ]:
# 3. for years 1980 to 1985
df_can.loc['Japan', [1980, 1981, 1982, 1983, 1984, 1984]]

In [42]:
df_can.loc['Japan',[1980,1981,1982,1983,1984,1985]]

1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object

In [ ]:
# Alternative Method
df_can.iloc[87, [3, 4, 5, 6, 7, 8]]

In [43]:
df_can.iloc[87,[3,4,5,6,7,8]]

1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object

**Exercise:** Let's view the number of immigrants from **Haiti** for the following scenarios: <br>1. The full row data (all columns) <br>2. For year 2000 <br>3. For years 1990 to 1995


In [46]:
df_can.loc['Haiti']
df_can.loc['Haiti',2000]
df_can.loc['Haiti',[1990,1991,1992,1993,1994,1995]]

1990    2379
1991    2829
1992    2399
1993    3655
1994    2100
1995    2014
Name: Haiti, dtype: object

In [ ]:
df_can.loc['Haiti']
df_can.loc['Haiti', 2000]
df_can.loc['Haiti', [1990, 1991, 1992, 1993, 1994, 1995]]

<details><summary>Click here for a sample python solution</summary>

```python
   # 1. the full row data (all columns)
    df_can.loc['Haiti']
    #or
    df_can[df_can.index == 'Haiti']

    # 2. for year 2000
    df_can.loc['Haiti', 2000]
           
    #  3. for years 1990 to 1995
    df_can.loc['Haiti', [1990, 1991, 1992, 1993, 1994, 1995]]
 ```

</details>


Column names that are integers (such as the years) might introduce some confusion. For example, when we are referencing the year 2013, one might confuse that when the 2013th positional index. 

To avoid this ambuigity, let's convert the column names into strings: '1980' to '2013'.


In [50]:
df_can.columns=list(map(str,df_can.columns))

In [49]:
df_can.columns = list(map(str, df_can.columns))
# [print (type(x)) for x in df_can.columns.values] #<-- uncomment to check type of column headers

Since we converted the years to string, let's declare a variable that will allow us to easily call upon the full range of years:


In [54]:
years=list(map(str,range(1980,2014)))
years

['1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013']

In [51]:
# useful for plotting later on
years = list(map(str, range(1980, 2014)))
years

['1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013']

**Exercise:** Create a list named 'year' using map function for years ranging from 1990 to 2013. <br>Then extract the data series from the dataframe df_can for Haiti using year list. 


In [56]:
year=list(map(str,range(1990,2014)))
df_can.loc['Haiti',year]

1990    2379
1991    2829
1992    2399
1993    3655
1994    2100
1995    2014
1996    1955
1997    1645
1998    1295
1999    1439
2000    1631
2001    2433
2002    2174
2003    1930
2004    1652
2005    1682
2006    1619
2007    1598
2008    2491
2009    2080
2010    4744
2011    6503
2012    5868
2013    4152
Name: Haiti, dtype: object

<details><summary>Click here for a sample python solution</summary>

```python
    #The correct answer is:
    year = list(map(str, range(1990, 2014)))
    haiti = df_can.loc['Haiti', year] # passing in years 1990 - 2013
    
```
</details>


### Filtering based on a criteria <a id="4"></a>
To filter the dataframe based on a condition, we simply pass the condition as a boolean vector. 

For example, Let's filter the dataframe to show the data on Asian countries (AreaName = Asia).


In [63]:
df_can[df_can['Continent']=='Asia']

Continent              Region  \
Afghanistan                                         Asia       Southern Asia   
Armenia                                             Asia        Western Asia   
Azerbaijan                                          Asia        Western Asia   
Bahrain                                             Asia        Western Asia   
Bangladesh                                          Asia       Southern Asia   
Bhutan                                              Asia       Southern Asia   
Brunei Darussalam                                   Asia  South-Eastern Asia   
Cambodia                                            Asia  South-Eastern Asia   
China                                               Asia        Eastern Asia   
China, Hong Kong Special Administrative Region      Asia        Eastern Asia   
China, Macao Special Administrative Region          Asia        Eastern Asia   
Cyprus                                              Asia        Western Asia   
Democratic People's Republic of Korea               Asia        Eastern Asia   
Georgia                                             Asia        Western Asia   
India                                               Asia       Southern Asia   
Indonesia                                           Asia  South-Eastern Asia   
Iran (Islamic Republic of)                          Asia       Southern Asia   
Iraq                                                Asia        Western Asia   
Israel                                              Asia        Western Asia   
Japan                                               Asia        Eastern Asia   
Jordan                                              Asia        Western Asia   
Kazakhstan                                          Asia        Central Asia   
Kuwait                                              Asia        Western Asia   
Kyrgyzstan                                          Asia        Central Asia   
Lao People's Democratic Republic                    Asia  South-Eastern Asia   
Lebanon                                             Asia        Western Asia   
Malaysia                                            Asia  South-Eastern Asia   
Maldives                                            Asia       Southern Asia   
Mongolia                                            Asia        Eastern Asia   
Myanmar                                             Asia  South-Eastern Asia   
Nepal                                               Asia       Southern Asia   
Oman                                                Asia        Western Asia   
Pakistan                                            Asia       Southern Asia   
Philippines                                         Asia  South-Eastern Asia   
Qatar                                               Asia        Western Asia   
Republic of Korea                                   Asia        Eastern Asia   
Saudi Arabia                                        Asia        Western Asia   
Singapore                                           Asia  South-Eastern Asia   
Sri Lanka                                           Asia       Southern Asia   
State of Palestine                                  Asia        Western Asia   
Syrian Arab Republic                                Asia        Western Asia   
Tajikistan                                          Asia        Central Asia   
Thailand                                            Asia  South-Eastern Asia   
Turkey                                              Asia        Western Asia   
Turkmenistan                                        Asia        Central Asia   
United Arab Emirates                                Asia        Western Asia   
Uzbekistan                                          Asia        Central Asia   
Viet Nam                                            Asia  South-Eastern Asia   
Yemen                                               Asia        Western Asia   

                                              

In [64]:
# 1. create the condition boolean series
condition = df_can['Continent'] == 'Asia'
print(condition)

Afghanistan                                              True
Albania                                                 False
Algeria                                                 False
American Samoa                                          False
Andorra                                                 False
Angola                                                  False
Antigua and Barbuda                                     False
Argentina                                               False
Armenia                                                  True
Australia                                               False
Austria                                                 False
Azerbaijan                                               True
Bahamas                                                 False
Bahrain                                                  True
Bangladesh                                               True
Barbados                                                False
Belarus 

In [65]:
# 2. pass this condition into the dataFrame
df_can[condition]

Continent              Region  \
Afghanistan                                         Asia       Southern Asia   
Armenia                                             Asia        Western Asia   
Azerbaijan                                          Asia        Western Asia   
Bahrain                                             Asia        Western Asia   
Bangladesh                                          Asia       Southern Asia   
Bhutan                                              Asia       Southern Asia   
Brunei Darussalam                                   Asia  South-Eastern Asia   
Cambodia                                            Asia  South-Eastern Asia   
China                                               Asia        Eastern Asia   
China, Hong Kong Special Administrative Region      Asia        Eastern Asia   
China, Macao Special Administrative Region          Asia        Eastern Asia   
Cyprus                                              Asia        Western Asia   
Democratic People's Republic of Korea               Asia        Eastern Asia   
Georgia                                             Asia        Western Asia   
India                                               Asia       Southern Asia   
Indonesia                                           Asia  South-Eastern Asia   
Iran (Islamic Republic of)                          Asia       Southern Asia   
Iraq                                                Asia        Western Asia   
Israel                                              Asia        Western Asia   
Japan                                               Asia        Eastern Asia   
Jordan                                              Asia        Western Asia   
Kazakhstan                                          Asia        Central Asia   
Kuwait                                              Asia        Western Asia   
Kyrgyzstan                                          Asia        Central Asia   
Lao People's Democratic Republic                    Asia  South-Eastern Asia   
Lebanon                                             Asia        Western Asia   
Malaysia                                            Asia  South-Eastern Asia   
Maldives                                            Asia       Southern Asia   
Mongolia                                            Asia        Eastern Asia   
Myanmar                                             Asia  South-Eastern Asia   
Nepal                                               Asia       Southern Asia   
Oman                                                Asia        Western Asia   
Pakistan                                            Asia       Southern Asia   
Philippines                                         Asia  South-Eastern Asia   
Qatar                                               Asia        Western Asia   
Republic of Korea                                   Asia        Eastern Asia   
Saudi Arabia                                        Asia        Western Asia   
Singapore                                           Asia  South-Eastern Asia   
Sri Lanka                                           Asia       Southern Asia   
State of Palestine                                  Asia        Western Asia   
Syrian Arab Republic                                Asia        Western Asia   
Tajikistan                                          Asia        Central Asia   
Thailand                                            Asia  South-Eastern Asia   
Turkey                                              Asia        Western Asia   
Turkmenistan                                        Asia        Central Asia   
United Arab Emirates                                Asia        Western Asia   
Uzbekistan                                          Asia        Central Asia   
Viet Nam                                            Asia  South-Eastern Asia   
Yemen                                               Asia        Western Asia   

                                              

In [ ]:
# we can pass multiple criteria in the same line.
# let's filter for AreaNAme = Asia and RegName = Southern Asia

df_can[(df_can['Continent']=='Asia') & (df_can['Region']=='Southern Asia')]

# note: When using 'and' and 'or' operators, pandas requires we use '&' and '|' instead of 'and' and 'or'
# don't forget to enclose the two conditions in parentheses

In [72]:
 # let's filter for AreaNAme = Asia and RegName = Southern Asia

df_can[(df_can['Continent']=='Asia') & (df_can['Region']=='Southern Asia')]

Continent         Region             DevName  1980  \
Afghanistan                     Asia  Southern Asia  Developing regions    16   
Bangladesh                      Asia  Southern Asia  Developing regions    83   
Bhutan                          Asia  Southern Asia  Developing regions     0   
India                           Asia  Southern Asia  Developing regions  8880   
Iran (Islamic Republic of)      Asia  Southern Asia  Developing regions  1172   
Maldives                        Asia  Southern Asia  Developing regions     0   
Nepal                           Asia  Southern Asia  Developing regions     1   
Pakistan                        Asia  Southern Asia  Developing regions   978   
Sri Lanka                       Asia  Southern Asia  Developing regions   185   

                            1981  1982  1983  1984  1985  1986   1987   1988  \
Afghanistan                   39    39    47    71   340   496    741    828   
Bangladesh                    84    86    81    98    92   486    503    476   
Bhutan                         0     0     0     1     0     0      0      0   
India                       8670  8147  7338  5704  4211  7150  10189  11522   
Iran (Islamic Republic of)  1429  1822  1592  1977  1648  1794   2989   3273   
Maldives                       0     0     1     0     0     0      0      0   
Nepal                          1     6     1     2     4    13      6     13   
Pakistan                     972  1201   900   668   514   691   1072   1334   
Sri Lanka                    371   290   197  1086   845  1838   4447   2779   

                             1989   1990   1991   1992   1993   1994   1995  \
Afghanistan                  1076   1028   1378   1170    713    858   1537   
Bangladesh                    387    611   1115   1655   1280   1361   2042   
Bhutan                          1      0      2      2      1      1      4   
India                       10343  12041  13734  13673  21496  18620  18489   
Iran (Islamic Republic of)   3781   3655   6250   6814   3959   2785   3956   
Maldives                        0      0      0      3      3      0      0   
Nepal                           4     23     29     32     40     31     66   
Pakistan                     2261   2470   3079   4071   4777   4666   4994   
Sri Lanka                    2758   3525   7266  13102   9563   7150   9368   

                             1996   1997   1998   1999   2000   2001   2002  \
Afghanistan                  2212   2555   1999   2395   3326   4067   3697   
Bangladesh                   2824   3378   2202   2064   3119   3831   2944   
Bhutan                          2      2      1      3      6      6      8   
India                       23859  22268  17241  18974  28572  31223  31889   
Iran (Islamic Republic of)   6205   7982   7057   6208   5884   6169   8129   
Maldives                        0      0      1      0      1      0      1   
Nepal                         132    155    104    157    236    272    363   
Pakistan                     9125  13073   9068   9979  15400  16708  15110   
Sri Lanka                    6484   5415   3566   4982   6081   5861   5279   

                             2003   2004   2005   2006   2007   2008   2009  \
Afghanistan                  3479   2978   3436   3009   2652   2111   1746   
Bangladesh                   2137   2660   4171   4014   2897   2939   2104   
Bhutan                          7      1      5     10      7     36    865   
India                       27155  28235  36210  33848  28742  28261  29456   
Iran (Islamic Republic of)   5918   6348   5837   7480   6974   6475   6580   
Maldives                        0      1      0      0      2      1      7   
Nepal                         313    404    607    540    511    581    561   
Pakistan                    13205  13399  14314  13127  10124   8994   7217   
Sri Lanka                    4892   4495   4930   4714   4123   4756   4547   

                             2010   2011  

**Exercise:** Fetch the data where AreaName is 'Africa' and RegName is 'Southern Africa'. <br>Display the dataframe and find out how many instances are there?


In [75]:
df_can[(df_can['Continent']=='Africa') & (df_can['Region']=='Southern Africa')]

Continent           Region             DevName  1980  1981  1982  \
Botswana        Africa  Southern Africa  Developing regions    10     1     3   
Lesotho         Africa  Southern Africa  Developing regions     1     1     1   
Namibia         Africa  Southern Africa  Developing regions     0     5     5   
South Africa    Africa  Southern Africa  Developing regions  1026  1118   781   
Swaziland       Africa  Southern Africa  Developing regions     4     1     1   

              1983  1984  1985  1986  1987  1988  1989  1990  1991  1992  \
Botswana         3     7     4     2     3     8     8     2     2     4   
Lesotho          2     7     5     3     1     6     5     1     4     4   
Namibia          3     2     1     1     4     3     2     2     4     0   
South Africa   379   271   310   718  1470  1305  1301   902   823  1059   
Swaziland        0    10     7     1     2     7     3     4     1     4   

              1993  1994  1995  1996  1997  1998  1999  2000  2001  2002  \
Botswana         4     0     2     3     2     0     2     3     2     4   
Lesotho          7     2     3     9     2     1     4     1     0     2   
Namibia          5     0     9     2     0     0     4     9     9    10   
South Africa  1549  2366  1455  1308  1729  1350  1478  1763  1899  1498   
Swaziland        2     3     5     2     4     3     4     2     2     1   

              2003  2004  2005  2006  2007  2008  2009  2010  2011  2012  \
Botswana         4     9     7    11     8    28    15    42    53    64   
Lesotho          2     2     4     0     4     1     8     7     1     0   
Namibia          7     9     6    19    13    26    14    16    23    24   
South Africa  1245  1175   988  1111  1200  1123  1188  1238   959  1243   
Swaziland        0     8     7     7     5     6    10     3    13    17   

              2013  Total  
Botswana        76    792  
Lesotho          6    214  
Namibia         83    640  
South Africa  1240  81136  
Swaziland       39    376

<details><summary>Click here for a sample python solution</summary>

```python
    df_can[(df_can['Continent']=='Africa') & (df_can['Region']=='Southern Africa')]
```
</details>


### Sorting Values of a Dataframe or Series <a id="5"></a><br>
You can use the `sort_values()` function is used to sort a DataFrame or a Series based on one or more columns. <br>You to specify the column(s) by which you want to sort and the order (ascending or descending). Below is the syntax to use it:-<br><br>
```df.sort_values(col_name, axis=0, ascending=True, inplace=False, ignore_index=False)```<br><br>
col_nam - the column(s) to sort by. <br>
axis - axis along which to sort. 0 for sorting by rows (default) and 1 for sorting by columns.<br>
ascending - to sort in ascending order (True, default) or descending order (False).<br>
inplace - to perform the sorting operation in-place (True) or return a sorted copy (False, default).<br>
ignore_index - to reset the index after sorting (True) or keep the original index values (False, default).<br>


Let's sort out dataframe df_can on 'Total' column, in descending order to find out the top 5 countries that contributed the most to immigration to Canada. 


In [ ]:
df_can.sort_values(by='Total', ascending=False, axis=0, inplace=True)
top_5 = df_can.head(5)
top_5

In [77]:
df_can.sort_values(by='Total',ascending=False, axis=0, inplace=True)
top_5=df_can.head(5)
top_5

Continent  \
India                                                     Asia   
China                                                     Asia   
United Kingdom of Great Britain and Northern Ireland    Europe   
Philippines                                               Asia   
Pakistan                                                  Asia   

                                                                  Region  \
India                                                      Southern Asia   
China                                                       Eastern Asia   
United Kingdom of Great Britain and Northern Ireland     Northern Europe   
Philippines                                           South-Eastern Asia   
Pakistan                                                   Southern Asia   

                                                                 DevName  \
India                                                 Developing regions   
China                                                 Developing regions   
United Kingdom of Great Britain and Northern Ireland   Developed regions   
Philippines                                           Developing regions   
Pakistan                                              Developing regions   

                                                       1980   1981   1982  \
India                                                  8880   8670   8147   
China                                                  5123   6682   3308   
United Kingdom of Great Britain and Northern Ireland  22045  24796  20620   
Philippines                                            6051   5921   5249   
Pakistan                                                978    972   1201   

                                                       1983   1984  1985  \
India                                                  7338   5704  4211   
China                                                  1863   1527  1816   
United Kingdom of Great Britain and Northern Ireland  10015  10170  9564   
Philippines                                            4562   3801  3150   
Pakistan                                                900    668   514   

                                                      1986   1987   1988  \
India                                                 7150  10189  11522   
China                                                 1960   2643   2758   
United Kingdom of Great Britain and Northern Ireland  9470  21337  27359   
Philippines                                           4166   7360   8639   
Pakistan                                               691   1072   1334   

                                                       1989   1990   1991  \
India                                                 10343  12041  13734   
China                                                  4323   8076  14255   
United Kingdom of Great Britain and Northern Ireland  23795  31668  23380   
Philippines                                           11865  12509  12718   
Pakistan                                               2261   2470   3079   

                                                       1992   1993   1994  \
India                                                 13673  21496  18620   
China                                                 10846   9817  13128   
United Kingdom of Great Britain and Northern Ireland  34123  33720  39231   
Philippines                                           13670  20479  19532   
Pakistan                                               4071   4777   4666   

                                                       1995   1996   1997  \
India                                                 18489  23859  22268   
China                                                 14398  19415  20475   
United Kingdom of Great Britain and Northern Ireland  30145  29322  22965   
Philippines                                           15864  13692  11549   
Pakistan                                              

**Exercise:** Find out top 3 countries that contributes the most to immigration to Canda in the year 2010. <br> Display the country names with the immigrant count in this year


In [78]:
df_can.sort_values(by='2010',inplace=True,ascending=False,axis=0)
top_3_2010=df_can.head(3)
top_3_2010

Continent              Region             DevName  1980  1981  \
Philippines      Asia  South-Eastern Asia  Developing regions  6051  5921   
India            Asia       Southern Asia  Developing regions  8880  8670   
China            Asia        Eastern Asia  Developing regions  5123  6682   

             1982  1983  1984  1985  1986   1987   1988   1989   1990   1991  \
Philippines  5249  4562  3801  3150  4166   7360   8639  11865  12509  12718   
India        8147  7338  5704  4211  7150  10189  11522  10343  12041  13734   
China        3308  1863  1527  1816  1960   2643   2758   4323   8076  14255   

              1992   1993   1994   1995   1996   1997   1998   1999   2000  \
Philippines  13670  20479  19532  15864  13692  11549   8735   9734  10763   
India        13673  21496  18620  18489  23859  22268  17241  18974  28572   
China        10846   9817  13128  14398  19415  20475  21049  30069  35529   

              2001   2002   2003   2004   2005   2006   2007   2008   2009  \
Philippines  13836  11707  12758  14004  18139  18400  19837  24887  28573   
India        31223  31889  27155  28235  36210  33848  28742  28261  29456   
China        36434  31961  36439  36619  42584  33518  27642  30037  29622   

              2010   2011   2012   2013    Total  
Philippines  38617  36765  34315  29544  1022782  
India        34235  27509  30933  33087  1383808  
China        30391  28502  33024  34129  1319924

<details><summary>Click here for a sample python solution</summary>

```python
    df_can.sort_values(by='2010', ascending=False, axis=0, inplace=True)
    top3_2010 = df_can['2010'].head(3)
    top3_2010
```
</details>


Congratulations! you have learned how to wrangle data with Pandas. You will be using alot of these commands to preprocess the data before its can be used for data visualization.


### Thank you for completing this lab!


## Author

<a href="https://www.linkedin.com/in/aklson/" target="_blank">Alex Aklson</a>


### Other Contributors
[Jay Rajasekharan](https://www.linkedin.com/in/jayrajasekharan),
[Ehsan M. Kermani](https://www.linkedin.com/in/ehsanmkermani),
[Slobodan Markovic](https://www.linkedin.com/in/slobodan-markovic),
[Weiqing Wang](https://www.linkedin.com/in/weiqing-wang-641640133/),
[Dr. Pooja](https://www.linkedin.com/in/p-b28802262/)

<!--   --!>
## Change Log


|  Date (YYYY-MM-DD) | Version | Changed By    |  Change Description                   |
|--------------------|---------|---------------|---------------------------------------|
| 2023-06-08         | 2.5     | Dr. Pooja         |  Separated from original lab        |
| 2021-05-29         | 2.4     | Weiqing Wang  |  Fixed typos and code smells.         |
| 2021-01-20         | 2.3     | Lakshmi Holla |  Changed TOC cell markdown            |
| 2020-11-20         | 2.2     | Lakshmi Holla |  Changed IBM box URL                  |
| 2020-11-03         | 2.1     | Lakshmi Holla |  Changed URL and info method          |
| 2020-08-27         | 2.0     | Lavanya       |  Moved Lab to course repo in GitLab   |--!>




## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
